govcd/api_vcd_versions.go

```diff
@@ -0,0 +1,192 @@
+func (vdcCli *VCDClient) APIClientVersionIs(versionConstraint string) bool {
+	err := vdcCli.vcdFetchSupportedVersions()
+	if err != nil {
+		util.Logger.Printf("[ERROR] could not retrieve supported versions: %s", err)
+		return false
+	}
+
+	util.Logger.Printf("[TRACE] checking current API version against constraints '%s'", versionConstraint)
+
+	isSupported, err := vdcCli.apiVerMatchesConstraint(vdcCli.Client.APIVersion, versionConstraint)
+	if err != nil {
+		util.Logger.Printf("[ERROR] unable to find cur supported version : %s", err)
+		return false
+	}
+
+	return isSupported
+}
+
+// vcdFetchSupportedVersions retrieves list of supported versions from
+// /api/versions endpoint and stores them in VCDClient for future uses.
+// It only does it once.
+func (vdcCli *VCDClient) vcdFetchSupportedVersions() error {
+	// Only fetch /versions if it is not stored already
+	numVersions := len(vdcCli.supportedVersions.VersionInfos)
+	if numVersions > 0 {
+		util.Logger.Printf("[TRACE] skipping fetch of versions because %d are stored", numVersions)
+		return nil
+	}
+
+	apiEndpoint := vdcCli.Client.VCDHREF
+	apiEndpoint.Path += "/versions"
+
+	req := vdcCli.Client.NewRequest(map[string]string{}, "GET", apiEndpoint, nil)
+	resp, err := checkResp(vdcCli.Client.Http.Do(req))
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	suppVersions := new(SupportedVersions)
+	err = decodeBody(resp, suppVersions)
+	if err != nil {
+		return fmt.Errorf("error decoding versions response: %s", err)
+	}
+
+	vdcCli.supportedVersions = *suppVersions
+
+	return nil
+}
+
+// maxSupportedVersion parses supported version list and returns the highest version in string format.
+func (vdcCli *VCDClient) maxSupportedVersion() (string, error) {
+	versions := make([]*semver.Version, len(vdcCli.supportedVersions.VersionInfos))
+	for i, raw := range vdcCli.supportedVersions.VersionInfos {
+		v, _ := semver.NewVersion(raw.Version)
+		versions[i] = v
+	}
+	// Sort supported versions in order lowest-highest
+	sort.Sort(semver.Collection(versions))
+
+	switch {
+	case len(versions) > 1:
+		return versions[len(versions)-1].Original(), nil
+	case len(versions) == 1:
+		return versions[0].Original(), nil
+	default:
+		return "", fmt.Errorf("could not identify supported versions")
+	}
+}
+
+// vcdCheckSupportedVersion checks if there is at least one specified version exactly matching listed ones.
+// Format example "27.0"
+func (vdcCli *VCDClient) vcdCheckSupportedVersion(version string) (bool, error) {
+	return vdcCli.checkSupportedVersionConstraint(fmt.Sprintf("= %s", version))
+}
+
+// Checks if there is at least one specified version matching the list returned by vCD.
+// Constraint format can be in format ">= 27.0, < 32",">= 30" ,"= 27.0".
+func (vdcCli *VCDClient) checkSupportedVersionConstraint(versionConstraint string) (bool, error) {
+	for _, vi := range vdcCli.supportedVersions.VersionInfos {
```

Comment:
what this `vi` is?

Reply:
single 'version info'. Like 
`for 'version info' in range vdcCli.supportedVersions.VersionInfos`

Reply:
Would be nice if name would be explicit :)